### PR TITLE
Normalize SHAMap visit functions

### PR DESCRIPTION
**Pull request by miguelportilla**
Normalized the SHAMap visit function parameter so that its return value is evaluated coherently.
visitDifference was made public.

### CODE REVIEW

File: src/ripple/app/misc/SHAMapStoreImp.cpp [hunk 1]
```diff
@@ -282,10 +282,10 @@ SHAMapStoreImp::copyNode (std::uint64_t& nodeCount,
     if (! (++nodeCount % checkHealthInterval_))
     {
         if (health())
```
Comment (bachase) [inline src/ripple/app/misc/SHAMapStoreImp.cpp]: Even if its private, is it worth documenting the return type in the header?

File: src/ripple/shamap/SHAMap.h [hunk 2]
```diff
@@ -177,10 +177,29 @@ class SHAMap
...
+    */
+    void visitNodes (std::function<bool (
+        SHAMapAbstractNode&)> const& function) const;
+
+    /**  Visit every node in this SHAMap that
+         is not present in the specified SHAMap
+
+         @param function called with every node visited.
+         If function returns false, visitDifferences exits.
+    */
```
Comment (bachase) [inline src/ripple/shamap/SHAMap.h]: Missing doxygen comment for the `have` param

File: src/ripple/shamap/impl/SHAMapSync.cpp [hunk 3]
```diff
@@ -100,6 +102,74 @@ void SHAMap::visitNodes(std::function<bool (SHAMapAbstractNode&)> const& functio
...
+void
+SHAMap::visitDifferences(SHAMap const* have,
+    std::function<bool (SHAMapAbstractNode&)> function) const
+{
+    // Visit every node in this SHAMap that is not present
+    // in the specified SHAMap
+    assert (root_->isValid ());
+
+    if (! root_)
+        return;
```
Comment (HowardHinnant) [inline src/ripple/shamap/impl/SHAMapSync.cpp]: Over time I believe that we have changed to never allow `root_ == nullptr`.  That being said, I have no problem with this check.  But shouldn't this come prior to the `assert (root_->isValid ());`?  If `root_ == nullptr`, and asserts are enabled, we're going to crash in the assert before we can return.

File: src/ripple/shamap/SHAMap.h [hunk 4]
```diff
@@ -177,10 +177,29 @@ class SHAMap
...
+    void visitDifferences(SHAMap const* have,
+        std::function<bool (SHAMapAbstractNode&)>) const;
```
Comment (HowardHinnant) [inline src/ripple/shamap/SHAMap.h]: Why was `visitDifferences` made public?  I don't see it used outside of `SHAMap::getFetchPack`.

File: src/ripple/shamap/impl/SHAMapSync.cpp (hunk 3 above)
Comment (HowardHinnant) [inline src/ripple/shamap/impl/SHAMapSync.cpp]: Hmm... I see we that this issue already existing here:  https://github.com/ripple/rippled/blob/develop/src/ripple/shamap/impl/SHAMapSync.cpp#L43-L46
Perhaps they should be reversed in both places.

File: src/ripple/shamap/SHAMap.h (hunk 4 above)
Comment (JoelKatz) [inline src/ripple/shamap/SHAMap.h]: The forthcoming sharding code needs `visitDifferences` to know which entries in a ledger to write to the shard versus which are already guaranteed to be there.

Comment (HowardHinnant) [inline src/ripple/shamap/SHAMap.h]: Ok, thanks.

File: src/ripple/app/misc/SHAMapStoreImp.cpp (hunk 1 above)
Comment (miguelportilla) [inline src/ripple/app/misc/SHAMapStoreImp.cpp]: Done

File: src/ripple/shamap/SHAMap.h (hunk 2 above)
Comment (miguelportilla) [inline src/ripple/shamap/SHAMap.h]: Fixed, thanks!

File: src/ripple/shamap/impl/SHAMapSync.cpp (hunk 3 above)
Comment (miguelportilla) [inline src/ripple/shamap/impl/SHAMapSync.cpp]: Nice catch!

**Pull request closed without merging**